mapviz_plugins/src/move_base_plugin.cpp

```diff
@@ -0,0 +1,338 @@
+{
+    if (is_mouse_down_)
+    {
+        QPointF head_pos = map_canvas_->MapGlCoordToFixedFrame( event->localPos() );
+        arrow_angle_ = atan2( head_pos.y() - arrow_tail_position_.y(),
+                              head_pos.x() - arrow_tail_position_.x() );
+    }
+    return false;
+}
+
+bool MoveBasePlugin::handleMouseRelease(QMouseEvent* event)
+{
+    if( !is_mouse_down_ )
+    {
+        return false;
+    }
+
+    is_mouse_down_ = false;
+
+    bool init_checked = ui_.pushButtonInitialPose->isChecked();
+    bool goal_checked = ui_.pushButtonGoalPose->isChecked();
+    if( !init_checked && !goal_checked)
+    {
+        return false;
+    }
+
+    tf::Quaternion quat = tf::createQuaternionFromYaw(arrow_angle_);
+
+    if( goal_checked ){
+
+        move_base_msg_.action_goal.header.frame_id = target_frame_;
+        move_base_msg_.action_goal.header.stamp = ros::Time::now();
+        move_base_msg_.action_goal.goal_id.stamp = move_base_msg_.action_goal.header.stamp;
+        move_base_msg_.action_goal.goal_id.id = "mapviz_goal";
+        move_base_msg_.action_goal.goal.target_pose.header = move_base_msg_.action_goal.header;
+
+        geometry_msgs::Pose& pose = move_base_msg_.action_goal.goal.target_pose.pose;
+        pose.position.x = arrow_tail_position_.x();
+        pose.position.y = arrow_tail_position_.y();
+        pose.position.z = 0.0;
+        tf::quaternionTFToMsg( quat, pose.orientation );
+
+        move_base_client_.sendGoal(move_base_msg_.action_goal.goal);
+        ui_.pushButtonGoalPose->setChecked(false);
+        monitoring_action_state_ = true;
+    }
+    if( init_checked ){
+        geometry_msgs::PoseWithCovarianceStamped initpose;
+        initpose.header.frame_id = target_frame_;
+        initpose.header.stamp = ros::Time::now();
+        initpose.pose.pose.position.x = arrow_tail_position_.x();
+        initpose.pose.pose.position.y = arrow_tail_position_.y();
+        initpose.pose.pose.position.z = 0.0;
+        tf::quaternionTFToMsg( quat, initpose.pose.pose.orientation );
+
+        init_pose_pub_.publish(initpose);
+        ui_.pushButtonInitialPose->setChecked(false);
+    }
+    return true;
+}
+
+
+void MoveBasePlugin::Draw(double x, double y, double scale)
+{
+    const QPointF arrow_points[7] = {
+        {10,0}, {6, -2.5}, {6.5,-1}, {0,-1}, {0,1}, {6.5, 1}, {6,2.5}
+    };
+
+    if( is_mouse_down_ )
+    {
+        QPointF transformed_points[7];
+        for (size_t i = 0; i<7; i++)
+        {
+            tf::Vector3 point(arrow_points[i].x(), arrow_points[i].y(), 0);
+            point *= scale*10;
+            point = tf::Transform( tf::createQuaternionFromYaw(arrow_angle_)  ) * point;
+            transformed_points[i] = QPointF(point.x() + arrow_tail_position_.x(),
+                                            point.y() + arrow_tail_position_.y() );
+        }
+        glColor3f(0.1, 0.9, 0.1);
+        glBegin(GL_TRIANGLE_FAN);
+        for (size_t i = 0; i<7; i++)
```

Comment:
It'd be preferable to use a range-based `for` here (like `for (const auto& point : transformed_points) { ... }`).

Reply:
you read my mind. I love range based for, but I noticed that the rest of the code don't use it much.

Reply:
Many of the plugins were written before Mapviz supported C++11, so they still use old-style for loops, but I prefer range-based loops for new code.